File: Graph/python/dfs.py
from typing import List

from .graph_utils import Graph


def depth_first_traversal(graph: Graph, source: str) -> List[str]:
    if not source:
        return []

    order: List[str] = []
    stack = [source]
    visited = set()

    while stack:
        node = stack.pop()
        if node in visited:
            continue
        visited.add(node)
        order.append(node)

        for neighbor in reversed(graph.get(node, [])):
            stack.append(neighbor)

    return order

File: Graph/php/dfs.php
<?php

declare(strict_types=1);

require_once __DIR__ . '/GraphUtils.php';

function depthFirstTraversal(array $graph, string $source): array
{
    if ($source === '') {
        return [];
    }

    $order = [];
    $stack = [$source];
    $visited = [];

    while ($stack) {
        $node = array_pop($stack);
        if (isset($visited[$node])) {
            continue;
        }
        $visited[$node] = true;
        $order[] = $node;

        $neighbors = $graph[$node] ?? [];
        for ($i = count($neighbors) - 1; $i >= 0; $i--) {
            $stack[] = $neighbors[$i];
        }
    }

    return $order;
}
